DataType.from_python does not work for Union and Optional types

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from typing import Optional, Union

def test_polars_from_python_type():
    python_type = int
    polars_type = pl.DataType.from_python(python_type)
    assert polars_type == pl.Int64


def test_polars_from_python_type_2():
    python_type = Optional[int]
    polars_type = pl.DataType.from_python(python_type)
    assert polars_type == pl.Int64


def test_polars_from_python_type_3():
    python_type = Union[int, None]
    polars_type = pl.DataType.from_python(python_type)
    assert polars_type == pl.Int64


def test_polars_from_python_type_4():
    python_type = int | None
    polars_type = pl.DataType.from_python(python_type)
    assert polars_type == pl.Int64

```


### Log output

```shell
FAILED tests/test_sql_attributes.py::test_polars_from_python_type_2 - TypeError: cannot parse input of type '_UnionGenericAlias' into Polars data type (given: typing.Optional[int])
FAILED tests/test_sql_attributes.py::test_polars_from_python_type_3 - TypeError: cannot parse input of type '_UnionGenericAlias' into Polars data type (given: typing.Optional[int])
``
```

### Issue description

`pl.DataType.from_python` does not work for `Optional` and `Union` types in version `1.37.1` (version `1.36.1` works as expected and all unit tests pass.

The problem is in `polars/datatypes/_parse.py::parse_into_dtype`(line 60) where the `isinstance(input, (UnionType, UnionTypeOld)` does not pass for `Union` and `Optional` types.

Replacing the condition `isinstance(input, (UnionType, UnionTypeOld)` with `isinstance(input, UnionType) or get_origin(input) is Union` seems to work in this case.

### Expected behavior

The expected behavior is for all the tests to pass - to detect `UnionType` properly in `parse_into_dtype` method.

### Installed versions

<details>

```
>>> pl.show_versions()
--------Version info---------
Polars:              1.37.1
Index type:          UInt32
Platform:            macOS-26.2-arm64-arm-64bit-Mach-O
Python:              3.13.7 (v3.13.7:bcee1c32211, Aug 14 2025, 19:10:51) [Clang 16.0.0 (clang-1600.0.26.6)]
Runtime:             rt32

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                1.42.33
cloudpickle          3.1.2
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.19.0
fsspec               <not installed>
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.10.8
numpy                1.26.4
openpyxl             3.1.5
pandas               3.0.0
polars_cloud         <not installed>
pyarrow              23.0.0
pydantic             2.12.5
pyiceberg            <not installed>
sqlalchemy           2.0.46
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

--- a/py-polars/src/polars/datatypes/_parse.py
+++ b/py-polars/src/polars/datatypes/_parse.py
@@ -7,7 +7,16 @@ from datetime import date, datetime, time, timedelta
 from decimal import Decimal as PyDecimal
 from inspect import isclass
 from types import NoneType, UnionType
-from typing import TYPE_CHECKING, Any, Final, ForwardRef, NoReturn, get_args
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Final,
+    ForwardRef,
+    NoReturn,
+    Union,
+    get_args,
+    get_origin,
+)
 
 import polars._reexport as pl
 from polars.datatypes.classes import (
@@ -33,9 +42,6 @@ if TYPE_CHECKING:
     from polars._typing import PolarsDataType, PythonDataType, SchemaDict
 
 
-UnionTypeOld = type(int | str)
-
-
 def parse_into_datatype_expr(input: Any) -> pl.DataTypeExpr:
     """Parse an input into a DataTypeExpr."""
     if isinstance(input, pl.DataTypeExpr):
@@ -57,7 +63,7 @@ def parse_into_dtype(input: Any) -> PolarsDataType:
         return input
     elif isinstance(input, ForwardRef):
         return _parse_forward_ref_into_dtype(input)
-    elif isinstance(input, (UnionType, UnionTypeOld)):
+    elif isinstance(input, UnionType) or get_origin(input) is Union:
         return _parse_union_type_into_dtype(input)
     else:
         return parse_py_type_into_dtype(input)

--- a/py-polars/tests/unit/datatypes/test_datatypes.py
+++ b/py-polars/tests/unit/datatypes/test_datatypes.py
@@ -3,7 +3,7 @@ from __future__ import annotations
 import copy
 import pickle
 from datetime import datetime, time, timedelta
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, Optional, Union
 
 import pytest
 
@@ -113,7 +113,6 @@ def test_dtypes_picklable() -> None:
 def test_dtypes_hashable() -> None:
     # ensure that all the types can be hashed, and that their hashes
     # are sufficient to ensure distinct entries in a dictionary/set
-
     all_dtypes = [
         getattr(datatypes, d)
         for d in dir(datatypes)
@@ -122,6 +121,15 @@ def test_dtypes_hashable() -> None:
     assert len(set(all_dtypes + all_dtypes)) == len(all_dtypes)
     assert len({pl.Datetime("ms"), pl.Datetime("us"), pl.Datetime("ns")}) == 3
     assert len({pl.List, pl.List(pl.Int16), pl.List(pl.Int32), pl.List(pl.Int64)}) == 4
+
+
+@pytest.mark.parametrize(
+    "python_type",
+    [int, int | None, Optional[int], Union[int, None]],  # noqa: UP007,UP045
+)
+def test_inference_from_python_type(python_type: Any) -> None:
+    polars_type = pl.DataType.from_python(python_type)
+    assert polars_type == pl.Int64
 
 
 @pytest.mark.parametrize(

